File: 2/1.py
from copy import deepcopy
import sys

lines = sys.stdin.read().split('\n')
n,m,k = list(map(int, lines[0].split()))
a = list(map(lambda l: list(map(int, l.split())), lines[1:]))
# n,m,k = list(map(int, input().split()))
# a = list(map(lambda line: list(map(int, line.split())), lines[1:]))
# a = []
# for i in range(n) :
    # a.append(list(map(int, input().split())))
vect = [0]*(k+1)
vect = [deepcopy(vect) for i in range(m)]
vect = [deepcopy(vect) for i in range(n)]

rightest = downest = True
for i in range(n-1,-1,-1) :
    downest = (i == n-1)
    for j in range(m-1,-1,-1) :
        rightest = (j == m-1)
        v = a[i][j]
        for q in range(k+1) :
            if rightest and downest :
                vect[i][j][q] = int((q==k) or (q+v==k))
            elif downest :
                vect[i][j][q] = vect[i][j+1][q]
                if q+v <= k :
                    vect[i][j][q] += vect[i][j+1][q+v]
            elif rightest :
                vect[i][j][q] = vect[i+1][j][q]
                if q+v <= k :
                    vect[i][j][q] += vect[i+1][j][q+v]
            else :
                vect[i][j][q] = vect[i][j+1][q] + vect[i+1][j][q]
                if q+v <= k :
                    vect[i][j][q] += vect[i][j+1][q+v] + vect[i+1][j][q+v]
            vect[i][j][q] %= (1e9+7)

print(int(vect[0][0][0]))

File: 2/1.cpp
#include <iostream>
using namespace std;
int main() {
    int n, m, k;
    cin >> n >> m >> k;
    int a[n][m];

    for(int i=0; i<n; i++)
        for(int j=0; j<m; j++)
            cin >> a[i][j];
    long int vect[m][k+1] = {0};
    
    for(int i=n-1; i>=0; i--) {
        bool downest = (i == n-1);
        for(int j=m-1; j>=0; j--) {
            bool rightest = (j == m-1);
            int v = a[i][j];
            for(int q=0; q<=k; q++) {
                if (rightest && downest) {
                    vect[j][q] = int((q==k) || (q+v==k));
                }
                else if (downest) {
                    vect[j][q] = vect[j+1][q];
                    if (q+v <= k)
                        vect[j][q] += vect[j+1][q+v];
                }
                else if (rightest) {
                    if (q+v <= k)
                        vect[j][q] += vect[j][q+v];
                }
                else {
                    vect[j][q] += vect[j+1][q];
                    if (q+v <= k)
                        vect[j][q] += vect[j+1][q+v] + vect[j][q+v];
                }
                vect[j][q] %= (1000000007);
            }
        }
    }
    cout << vect[0][0] << endl;
}